sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/LookupCatalog.scala

```diff
@@ -140,19 +140,22 @@ private[sql] trait LookupCatalog extends Logging {
    * For legacy support only. Please use [[CatalogAndIdentifier]] instead on DSv2 code paths.
    */
   object AsTableIdentifier {
-    def unapply(parts: Seq[String]): Option[TableIdentifier] = parts match {
-      case CatalogAndMultipartIdentifier(None, names)
+    def unapply(parts: Seq[String]): Option[TableIdentifier] = {
+      def namesToTableIdentifier(names: Seq[String]): Option[TableIdentifier] = names match {
+        case Seq(name) => Some(TableIdentifier(name))
+        case Seq(database, name) => Some(TableIdentifier(name, Some(database)))
+        case _ => None
+      }
+      parts match {
+        case CatalogAndMultipartIdentifier(None, names)
           if CatalogV2Util.isSessionCatalog(currentCatalog) =>
-        names match {
-          case Seq(name) =>
-            Some(TableIdentifier(name))
-          case Seq(database, name) =>
-            Some(TableIdentifier(name, Some(database)))
-          case _ =>
-            None
-        }
-      case _ =>
-        None
+          namesToTableIdentifier(names)
+        case CatalogAndMultipartIdentifier(Some(catalog), names)
+          if CatalogV2Util.isSessionCatalog(catalog) &&
+             CatalogV2Util.isSessionCatalog(currentCatalog) =>
```

Comment:
if catalog name is specified, why do we care about the current catalog?

Reply:
So, `isSessionCatalog()` checks that the parameter is `spark_catalog`. If the specified catalog is `spark_catalog` but the current one **is not** `spark_catalog`, and we cut it off from names, then we "bind" table name to something different from `spark_catalog`, correct?

Reply:
no, `TableIdentifier` is only used in v1 commands. V1 commands can't look up catalogs and the current catalog doesn't matter to it.

Reply:
`AsTableIdentifier` is used directly in public API methods:
- https://github.com/apache/spark/blob/2da72593c1cf63fc6f815416b8d553f0a53f3e65/sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala#L495
- https://github.com/apache/spark/blob/2da72593c1cf63fc6f815416b8d553f0a53f3e65/sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala#L625

If I remove the check, this could impact on their behavior.